### ISSUE TITLE
Adding "unique: true" to existing @Column yields constraint with wrong name, throwing "relation ... already exists" on every restart

### ISSUE THREAD

**Issue by avbentem** [typeorm/typeorm#770]
A mismatch in the naming of `UNIQUE` constraints between newly created tables, and tables that have `unique: true` added to new or existing `@Column` definitions at a later time, causes errors like:

    error: relation "uk_my_column" already exists

When using `autoSchemaSync: true` in typeorm 0.1.0-alpha.34 along with pg 7.1.2 towards PostgreSQL 9.6.3 on x86_64-pc-linux-gnu, compiled by gcc (Debian 4.9.2-10) 4.9.2, 64-bit:

1. All is fine when creating a new table with a column which has `unique: true` specified for the very first run. For that, the following is used in the `CREATE TABLE` statement:

    ```sql
    CREATE TABLE ... CONSTRAINT "uk_my_table_my_column" UNIQUE ("my_column")
    ```

2. However, when **adding `unique: true` to an existing column**, TypeORM adds a constraint with a name that lacks the table name. The `ALTER TABLE ... ADD` uses:

    ```sql
    ALTER TABLE ... ADD CONSTRAINT "uk_my_column" UNIQUE ("my_column")
    ```
    This works fine initially (the constraint is added), but fails for subsequent restarts for which TypeORM tries to add the very same constraint again and again. So:

    **The erroneous shorter name seems to be the culprit, as TypeORM probably expects the long name when trying to determine if the unique constraint already exists**. 

3. And when **adding a column to an existing table and immediately specifying `unique: true` for that new column**, then the `UNIQUE` constraint is not added immediately. Instead, TypeORM adds the column on the first restart, and only adds the constraint on the second restart, with the wrong shorter name just like above, so then also failing for subsequent restarts.

I guess I could create a PR for (2), which would also partly fix (3), but the latter should obviously add the new constraint in the first `ALTER TABLE ... ADD COLUMN`.

- This might also be the cause for problems with `unique: true` in `@Index`, like reported in #721, but I am not sure. 
- This might also be the cause for #719, though that (erroneously?) refers to foreign keys. (And uses `schema:sync`, but that might work the same as `autoSchemaSync: true`.)

More details below, but I guess the points above cover it all.

## Adding `unique: true` to existing `@Column`

With:

```typescript
import {Column, PrimaryGeneratedColumn} from 'typeorm';
import {Entity} from 'typeorm/decorator/entity/Entity';

@Entity('my_table')
export class MyTable {
    @PrimaryGeneratedColumn()
    id: number;

    @Column({name: 'my_column'})
    myColumn: string;
}
```
...the table gets created as:

```sql
executing query: START TRANSACTION
executing query: CREATE SCHEMA IF NOT EXISTS "public"
creating a new table: my_table
executing query: CREATE TABLE "my_table" ("id" SERIAL NOT NULL, "my_column" character varying NOT NULL, PRIMARY KEY("id"))
executing query: COMMIT
```

Next, when changing to

```typescript
    @Column({name: 'my_column', unique: true})
```

...then with `logging: 'all'` I see:

```sql
executing query: SELECT * FROM information_schema.tables WHERE table_catalog = 'postgres' AND table_schema = 'public' AND table_name IN ('my_table', 'my_other_table')
executing query: SELECT * FROM information_schema.columns WHERE table_catalog = 'postgres' AND table_schema = 'public'
executing query: SELECT t.relname AS table_name, i.relname AS index_name, a.attname AS column_name  FROM pg_class t, pg_class i, pg_index ix, pg_attribute a, pg_namespace ns
WHERE t.oid = ix.indrelid AND i.oid = ix.indexrelid AND a.attrelid = t.oid
AND a.attnum = ANY(ix.indkey) AND t.relkind = 'r' AND t.relname IN ('my_table', 'my_other_table') AND t.relnamespace = ns.OID AND ns.nspname ='public' ORDER BY t.relname, i.relname
executing query: SELECT table_name, constraint_name FROM information_schema.table_constraints WHERE table_catalog = 'postgres' AND table_schema = 'public' AND constraint_type = 'FOREIGN KEY'
executing query: SELECT * FROM information_schema.table_constraints WHERE table_catalog = 'postgres' AND table_schema = 'public' AND constraint_type = 'UNIQUE'
executing query: SELECT c.column_name, tc.table_name, tc.constraint_name FROM information_schema.table_constraints tc
JOIN information_schema.constraint_column_usage AS ccu USING (constraint_schema, constraint_name)
JOIN information_schema.columns AS c ON c.table_schema = tc.constraint_schema AND tc.table_name = c.table_name AND ccu.column_name = c.column_name
where constraint_type = 'PRIMARY KEY' AND c.table_schema = 'public' and tc.table_catalog = 'postgres'

executing query: START TRANSACTION
executing query: CREATE SCHEMA IF NOT EXISTS "public"
columns changed in my_table. updating: my_column
executing query: ALTER TABLE "my_table" ADD CONSTRAINT "uk_my_column" UNIQUE ("my_column")
executing query: COMMIT
```

So, so far so good, the column constraint was added.

However, on subsequent restarts, without any further changes to the model, it shows the same initial queries, and then tries to add the same constraint again:

```sql
executing query: START TRANSACTION
executing query: CREATE SCHEMA IF NOT EXISTS "public"
columns changed in my_table. updating: my_column
executing query: ALTER TABLE "my_table" ADD CONSTRAINT "uk_my_column" UNIQUE ("my_column")
query failed: ALTER TABLE "my_table" ADD CONSTRAINT "uk_my_column" UNIQUE ("my_column")
error: { error: relation "uk_my_column" already exists
    at Connection.parseE (/Users/arjan/dev/workspaces/ipp/server/node_modules/pg/lib/connection.js:546:11)
    at Connection.parseMessage (/Users/arjan/dev/workspaces/ipp/server/node_modules/pg/lib/connection.js:371:19)
    at Socket.<anonymous> (/Users/arjan/dev/workspaces/ipp/server/node_modules/pg/lib/connection.js:114:22)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
  name: 'error',
  length: 99,
  severity: 'ERROR',
  code: '42P07',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'index.c',
  line: '793',
  routine: 'index_create' }
executing query: ROLLBACK
```

Note that `SELECT * FROM information_schema.table_constraints WHERE table_catalog = 'postgres' AND table_schema = 'public' AND constraint_type = 'UNIQUE'` nicely shows the existing `uk_my_column` for that very table.

## Using `@Column({unique: true})` when creating a new table

When adding the `unique: true` from the beginning, the table gets to be created using:

```sql
executing query: CREATE TABLE "my_table" ("id" SERIAL NOT NULL, "my_column" character varying NOT NULL, CONSTRAINT "uk_my_table_my_column" UNIQUE ("my_column"), PRIMARY KEY("id"))
```

## Adding `@Column({unique: true})` to existing table

When adding a new column including `unique: true`, the column is initially added without the constraint, using:

```sql
executing query: ALTER TABLE "my_table" ADD "my_new_column" character varying NOT NULL
```

**Comment by zegerhoogeboom** [comment]
+1 Just ran into this exact issue.

**Comment by pleerock** [comment]
Thank you for such detailed explanation. Im on vacation right now and will return back to opensource in September. Then I'll try to fix this issue.

**Comment by iz-iznogood** [comment]
This is fixed now but in order to work it needs some previous pull requests to be merged to master. When this happens I will send a pull request

**Comment by pleerock** [comment]
Fix provided by @iz-iznogood is released in `0.1.0-alpha.40`, please check a new version.
